src/core/sock/sock-redirect.cpp

```diff
@@ -369,27 +369,42 @@ int init_child_process_for_nginx()
             sockinfo_udp *udp_sock = dynamic_cast<sockinfo_udp *>(parent_sock_fd_api);
             if (udp_sock) {
-                int optval;
-                socklen_t optlen = sizeof(optval);
-                if (udp_sock->getsockopt(SOL_SOCKET, SO_REUSEPORT, &optval, &optlen) < 0) {
+                int reuse_port;
+                socklen_t optlen = sizeof(reuse_port);
+                if (udp_sock->getsockopt(SOL_SOCKET, SO_REUSEPORT, &reuse_port, &optlen) < 0) {
                     srdr_logdbg("fd=%d - getsockopt() failed", sock_fd);
                     continue;
                 }
                 uint16_t port = ntohs(sa.get_in_port());
-                bool use_as_udp_listen_socket = optval && port;
-                if (!use_as_udp_listen_socket) {
+                bool use_as_udp_listen_socket = false;
+                /*
+                 * Specific NGINX implementation.
+                 *
+                 * In case of "reuseport" directive
+                 * NGINX master process creates a UDP socket per worker process per port before it
+                 * forks. Therefore, each worker process attaches a single UDP socket out of
+                 * #worker_processes.
+                 *
+                 * Without "reuseport" directive, NGINX master process creates a single UDP socket
+                 * before it forks. Therefore, all worker processes attach the UDP socket (single).
+                 */
+
+                if (port) {
+                    if (reuse_port) {
+                        // In case of "reuseport", Nginx master process creates #worker_processes
+                        // number of sockets before fork. In order to use a single socket per
+                        // worker we choose a socket by g_worker_index;
+                        use_as_udp_listen_socket = (udp_sockets_per_port[port] == g_worker_index);
+                    } else {
+                        // Without "reuseport", Nginx master process creates a single socket before
+                        // fork. Use this socket for all workers - always true;
+                        use_as_udp_listen_socket = true;
+                    }
+                } else {
                     continue;
                 }
 
-                /*
-                 * DANGER: This code depends on a very specific NGINX implementation detail.
-                 * NGINX master process creates a QUIC UDP socket per worker process per port
-                 * before fork(). To ensure that the copy of the global g_p_fd_collection that
-                 * this child process has contains *only* the `sockinfo_udp` pointers that
-                 * belong to it we discard all the `sock_fd`s we aggregated in the
-                 * udp_sockets_per_port and take the one indexed by the g_worker_index.
-                 */
-                if (unlikely(udp_sockets_per_port[port] == g_worker_index)) {
+                if (use_as_udp_listen_socket) {
```

Comment:
could you consider refactoring
```
                uint16_t port = ntohs(sa.get_in_port());
                int reuse_port;
                socklen_t optlen = sizeof(reuse_port);
                if ((port == 0) || (udp_sock->getsockopt(SOL_SOCKET, SO_REUSEPORT, &reuse_port, &optlen) < 0)) {
                    continue;
                }
/* Description....
*/
if ((reuse_port == 0) || unlikely(udp_sockets_per_port[port] == g_worker_index)) {
...
```

Reply:
i kept use_as_udp_listen_socket variable so comments and logic will be clear for the future.
fixed the first comment.

Reply:
i am not sure that existing code more clear. In addition it uses at least two `if condition` all time instead of one.